Build system help

I'm on macOS 10.13.6 because of my hardware.
I'm getting an error when building Kitty.app about `UserNotifications.h` not being found.
The changelog for 0.19 mentions the new notification framework which seems like the likely culprit.
A little perusal of the source shows there's [an ifdef for people in a similar situation](https://github.com/kovidgoyal/kitty/blob/master/kitty/cocoa_window.m#L12).

Here comes the question. I can't find any documentation on how to specify this value. Do I set an env var or do I need to pass it to setup.py?

Oh, actually. The build is failing because of xmlsec, a dependency I guess?

I'm not sure if this is on kitty though, might just be nixpkgs has a bad version of xmlsec on unstable at the moment.

Truncated log because it's way too many similar blocks:
```
Error: failed to parse xml file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to load template "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to decrypt file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Test: /nist-aesgcm/aes256/aes256-gcm-96-408-0-128-14 (success)
/private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 check-transforms  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config aes256-gcm
Transforms "aes256-gcm" found
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 decrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-14.xml
=== TEST RESULT: 0; expected: success
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 encrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml --binary-data /tmp/testEnc.sh.20201018_194954-33598.tmp.3 --output /tmp/testEnc.sh.20201018_194954-33598.tmp /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-14.tmpl
func=xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock:file=ciphers.c:line=183:obj=unknown:subj=in != NULL:error=100:assertion: 
func=xmlSecOpenSSLEvpBlockCipherGCMCtxFinal:file=ciphers.c:line=523:obj=aes256-gcm:subj=xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock:error=1:xmlsec library function failed: 
func=xmlSecOpenSSLEvpBlockCipherExecute:file=ciphers.c:line=844:obj=aes256-gcm:subj=xmlSecOpenSSLEvpBlockCipherCtxFinal:error=1:xmlsec library function failed: 
func=xmlSecTransformDefaultPushBin:file=transforms.c:line=1929:obj=aes256-gcm:subj=xmlSecTransformExecute:error=1:xmlsec library function failed:final=1
func=xmlSecTransformPump:file=transforms.c:line=1414:obj=aes256-gcm:subj=xmlSecTransformPushBin:error=1:xmlsec library function failed: 
func=xmlSecTransformCtxUriExecute:file=transforms.c:line=997:obj=input-uri:subj=xmlSecTransformPump:error=1:xmlsec library function failed: 
func=xmlSecTransformCtxExecute:file=transforms.c:line=1101:obj=unknown:subj=xmlSecTransformCtxUriExecute:error=1:xmlsec library function failed: 
func=xmlSecEncCtxUriEncrypt:file=xmlenc.c:line=488:obj=unknown:subj=xmlSecTransformCtxExecute:error=1:xmlsec library function failed: 
Error: failed to encrypt file "/tmp/testEnc.sh.20201018_194954-33598.tmp.3"
Error: failed to encrypt file with template "/private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-14.tmpl"
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 decrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml --output /tmp/testEnc.sh.20201018_194954-33598.tmp.2 /tmp/testEnc.sh.20201018_194954-33598.tmp
I/O warning : failed to load external entity "/tmp/testEnc.sh.20201018_194954-33598.tmp"
func=xmlSecParseFile:file=parser.c:line=373:obj=unknown:subj=xmlCreateFileParserCtxt:error=5:libxml2 library function failed:filename=/tmp/testEnc.sh.20201018_194954-33598.tmp; xml error: 1549: failed to load external entity "/tmp/testEnc.sh.20201018_194954-33598.tmp"

Error: failed to parse xml file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to load template "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to decrypt file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Test: /nist-aesgcm/aes256/aes256-gcm-96-408-0-128-15 (success)
/private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 check-transforms  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config aes256-gcm
Transforms "aes256-gcm" found
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 decrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-15.xml
=== TEST RESULT: 0; expected: success
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 encrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml --binary-data /tmp/testEnc.sh.20201018_194954-33598.tmp.3 --output /tmp/testEnc.sh.20201018_194954-33598.tmp /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-15.tmpl
func=xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock:file=ciphers.c:line=183:obj=unknown:subj=in != NULL:error=100:assertion: 
func=xmlSecOpenSSLEvpBlockCipherGCMCtxFinal:file=ciphers.c:line=523:obj=aes256-gcm:subj=xmlSecOpenSSLEvpBlockCipherCtxUpdateBlock:error=1:xmlsec library function failed: 
func=xmlSecOpenSSLEvpBlockCipherExecute:file=ciphers.c:line=844:obj=aes256-gcm:subj=xmlSecOpenSSLEvpBlockCipherCtxFinal:error=1:xmlsec library function failed: 
func=xmlSecTransformDefaultPushBin:file=transforms.c:line=1929:obj=aes256-gcm:subj=xmlSecTransformExecute:error=1:xmlsec library function failed:final=1
func=xmlSecTransformPump:file=transforms.c:line=1414:obj=aes256-gcm:subj=xmlSecTransformPushBin:error=1:xmlsec library function failed: 
func=xmlSecTransformCtxUriExecute:file=transforms.c:line=997:obj=input-uri:subj=xmlSecTransformPump:error=1:xmlsec library function failed: 
func=xmlSecTransformCtxExecute:file=transforms.c:line=1101:obj=unknown:subj=xmlSecTransformCtxUriExecute:error=1:xmlsec library function failed: 
func=xmlSecEncCtxUriEncrypt:file=xmlenc.c:line=488:obj=unknown:subj=xmlSecTransformCtxExecute:error=1:xmlsec library function failed: 
Error: failed to encrypt file "/tmp/testEnc.sh.20201018_194954-33598.tmp.3"
Error: failed to encrypt file with template "/private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/aes256/aes256-gcm-96-408-0-128-15.tmpl"
 /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/apps/xmlsec1 decrypt  --crypto openssl --crypto-config /tmp/xmlsec-crypto-config --keys-file /private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30/tests/nist-aesgcm/keys-aes256-gcm.xml --output /tmp/testEnc.sh.20201018_194954-33598.tmp.2 /tmp/testEnc.sh.20201018_194954-33598.tmp
I/O warning : failed to load external entity "/tmp/testEnc.sh.20201018_194954-33598.tmp"
func=xmlSecParseFile:file=parser.c:line=373:obj=unknown:subj=xmlCreateFileParserCtxt:error=5:libxml2 library function failed:filename=/tmp/testEnc.sh.20201018_194954-33598.tmp; xml error: 1549: failed to load external entity "/tmp/testEnc.sh.20201018_194954-33598.tmp"

Error: failed to parse xml file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to load template "/tmp/testEnc.sh.20201018_194954-33598.tmp"
Error: failed to decrypt file "/tmp/testEnc.sh.20201018_194954-33598.tmp"
make[1]: *** [Makefile:1184: check-crypto-openssl] Error 188
make[1]: Leaving directory '/private/var/folders/xm/n96l21b93xx1bm16cjr16gq40000gn/T/nix-build-xmlsec-1.2.30.drv-0/xmlsec1-1.2.30'
make: *** [Makefile:1179: check-all] Error 1
builder for '/nix/store/vpi94fq88k2a5sws953n3c7s895k8gad-xmlsec-1.2.30.drv' failed with exit code 2
cannot build derivation '/nix/store/c848559pvhkvgkagyap3wz198b6620w6-oath-toolkit-2.6.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/bc995y2fw3isn7zbrz91qlwa5zyijfzb-pass-extensions-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/a2l5x4fq0i4ilj7qqblav32x5lj15fs4-home-manager-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/i5hwkrpqb3bphiq9mjmh7z7g98ib1jd5-pass-extensions-env-fish-completions.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/p3a8pdlgaybgjvlmq11fvk00dxvpzbrq-home-manager-generation.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/p62jpdyn640yg83f7aspvpi159iiw3j5-activation-toonn.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wb8vaq83zhv1x88jdbknd0861hjn7f6h-darwin-system-21.03pre246624.cfed29bfcb2+darwin4.0000000.drv': 1 dependencies couldn't be built
error: build of '/nix/store/wb8vaq83zhv1x88jdbknd0861hjn7f6h-darwin-system-21.03pre246624.cfed29bfcb2+darwin4.0000000.drv' failed
```

--- a/setup.py
+++ b/setup.py
@@ -343,7 +343,10 @@ def kitty_env() -> Env:
         platform_libs = [
             '-framework', 'CoreText', '-framework', 'CoreGraphics',
         ]
-        user_notifications_framework = first_successful_compile(ans.cc, '-framework UserNotifications')
+        test_program_src = '''#include <UserNotifications/UserNotifications.h>
+        int main(void) { return 0; }\n'''
+        user_notifications_framework = first_successful_compile(ans.cc, ' -framework UserNotifications',
+                                                                src=test_program_src, lang='objective-c')
         if user_notifications_framework:
             platform_libs.extend(shlex.split(user_notifications_framework))
         else:

--- a/setup.py
+++ b/setup.py
@@ -13,6 +13,7 @@ import shutil
 import subprocess
 import sys
 import sysconfig
+import tempfile
 import time
 from contextlib import suppress
 from functools import partial
@@ -211,25 +212,27 @@ def get_sanitize_args(cc: str, ccver: Tuple[int, int]) -> List[str]:
     return sanitize_args
 
 
-def test_compile(cc: str, *cflags: str, src: Optional[str] = None) -> bool:
+def test_compile(cc: str, *cflags: str, src: Optional[str] = None, lang: Optional[str] = None) -> bool:
+    lang = lang or 'c'
     src = src or 'int main(void) { return 0; }'
-    p = subprocess.Popen(
-        [cc] + list(cflags) + ['-x', 'c', '-o', os.devnull, '-'],
-        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, stdin=subprocess.PIPE
-    )
-    stdin = p.stdin
-    assert stdin is not None
-    try:
-        stdin.write(src.encode('utf-8'))
-        stdin.close()
-    except BrokenPipeError:
-        return False
-    return p.wait() == 0
+    with tempfile.TemporaryDirectory() as tdir:
+        p = subprocess.Popen(
+            [cc] + list(cflags) + ['-x', lang, '-o', tdir + '/result', '-'],
+            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, stdin=subprocess.PIPE
+        )
+        stdin = p.stdin
+        assert stdin is not None
+        try:
+            stdin.write(src.encode('utf-8'))
+            stdin.close()
+        except BrokenPipeError:
+            return False
+        return p.wait() == 0
 
 
-def first_successful_compile(cc: str, *cflags: str, src: Optional[str] = None) -> str:
+def first_successful_compile(cc: str, *cflags: str, src: Optional[str] = None, lang: Optional[str] = None) -> str:
     for x in cflags:
-        if test_compile(cc, *shlex.split(x), src=src):
+        if test_compile(cc, *shlex.split(x), src=src, lang=lang):
             return x
     return ''
 

